### PR TITLE
Delegate to subresource when validating and updating

### PR DESCRIPTION
**Pull request by tsx**

### CODE REVIEW

File: flask_mongorest/resources.py [hunk 1]
```diff
@@ -199,15 +199,22 @@ def serialize_field(self, obj, **kwargs):
         else:
             return self.serialize(obj, **kwargs)
 
+    def _needs_delegation(self, obj):
+        # If a subclass of an obj has been called with a base class' resource,
+        return obj.__class__ in self._child_document_resources \
+        and self._child_document_resources[obj.__class__] != self.__class__
+
+    def _create_subresource(self, obj):
+        return self._child_document_resources[obj.__class__]()
```
Comment (thomasst) [inline flask_mongorest/resources.py]: Can we automatically assign the `data` property to the subresource here?

File: flask_mongorest/resources.py [hunk 2]
```diff
@@ -199,15 +200,24 @@ def serialize_field(self, obj, **kwargs):
         else:
             return self.serialize(obj, **kwargs)
 
+    def _subresource_class(self, obj):
```
Comment (thomasst) [inline flask_mongorest/resources.py]: Even though this is an internal method, we should add a docstring because it's not immediately clear what it's doing (explain `default_child_resource_document` etc.).

File: flask_mongorest/resources.py [hunk 3]
```diff
@@ -199,15 +200,24 @@ def serialize_field(self, obj, **kwargs):
...
         if not obj:
             return {}
 
         # If a subclass of an obj has been called with a base class' resource,
         # use the subclass-specific serialization
-        if obj.__class__ in self._child_document_resources \
-        and self._child_document_resources[obj.__class__] != self.__class__:
-            return obj and self._child_document_resources[obj.__class__]().serialize_field(obj, **kwargs)
+        if self._subresource_class(obj):
+            return self._create_subresource(obj).serialize(obj, **kwargs)
```
Comment (thomasst) [inline flask_mongorest/resources.py]: We end up calling `self._subresource_class(obj)` twice here (once directly, once through `_create_subresource`). Ideally we would only call it once. Same in the other methods.

**Comment by philfreo** [comment]
@thomasst or @wojcikstefan pls review

File: flask_mongorest/resources.py [hunk 4]
```diff
@@ -199,15 +201,26 @@ def serialize_field(self, obj, **kwargs):
         else:
             return self.serialize(obj, **kwargs)
 
+    def _subresource(self, obj):
+        """ Selects and creates an appropriate sub-resource class for delegation or return None if there isn't one """
```
Comment (wojcikstefan) [inline flask_mongorest/resources.py]: According to https://www.python.org/dev/peps/pep-0257/#id16

> There's no blank line either before or after the docstring.

File: flask_mongorest/resources.py [hunk 5]
```diff
@@ -199,15 +201,26 @@ def serialize_field(self, obj, **kwargs):
...
+        default_s_class = self._child_document_resources[self.default_child_resource_document] if self.default_child_resource_document else None
+        s_class = self._child_document_resources.get(obj.__class__, default_s_class)
+        if s_class and s_class != self.__class__:
+            r = s_class()
+            r.data = self.data
+            return r
+        else:
+            return None
```
Comment (wojcikstefan) [inline flask_mongorest/resources.py]: Is there ever an expected case where a subresource isn't found? If not, this should raise an exception.

File: flask_mongorest/resources.py [hunk 6]
```diff
@@ -344,6 +357,14 @@ def value_for_field(self, obj, field):
         raise UnknownFieldError
 
     def validate_request(self, obj=None):
+        if (request.method == 'PUT' and obj) or request.method == 'POST':
```
Comment (wojcikstefan) [inline flask_mongorest/resources.py]: These conditions are always met - we only use `validate_request` in `ResourceView#post` and `ResourceView#put` (where for PUT, `obj` is never `None`).

Comment (tsx) [inline flask_mongorest/resources.py]: We have mass-updating when you specify a filter and PUT some fields (used in tasks). For those, validate_request is called once for request data on PUT

File: flask_mongorest/resources.py (hunk 5 above)
Comment (tsx) [inline flask_mongorest/resources.py]: Return value of `None` indicates that there is no subclass to delegate to (think simple, non-inherited resources). Existence of the result is a signal to caller to perform actual delegation.

File: flask_mongorest/resources.py (hunk 6 above)
Comment (wojcikstefan) [inline flask_mongorest/resources.py]: I thought that even bulk PUTs still call `validate_request(obj)` for each object in the bulk list.

**Comment by wojcikstefan** [comment]
@tsx should we add a unit test utilizing this logic?

File: flask_mongorest/resources.py (hunk 6 above)
Comment (tsx) [inline flask_mongorest/resources.py]: Well without this line it breaks tests in main repo (I actually ported that condition from the main repo, maybe that's some corner case we haven't accounted for)

Comment (wojcikstefan) [inline flask_mongorest/resources.py]: I'm curious what tests break and whether the main repo is utilizing this method incorrectly (or if my assumption that this method is supposed to be used only with a POST or a PUT *and* an `obj` is incorrect).

**Comment by philfreo** [comment]
@wojcikstefan pls review & merge if ready

File: flask_mongorest/resources.py [hunk 7]
```diff
@@ -199,15 +201,26 @@ def serialize_field(self, obj, **kwargs):
         else:
             return self.serialize(obj, **kwargs)
 
+    def _subresource(self, obj):
+        """Selects and creates an appropriate sub-resource class for delegation or return None if there isn't one"""
+        default_s_class = self._child_document_resources[self.default_child_resource_document] if self.default_child_resource_document else None
+        s_class = self._child_document_resources.get(obj.__class__, default_s_class)
```
Comment (wojcikstefan) [inline flask_mongorest/resources.py]: Assuming that `_child_document_resources[obj.__class__]` never returns a falsey value, the code below is equivalent and IMHO much cleaner. Do you agree @tsx?

```
s_class = self._child_document_resources.get(obj.__class__)
if not s_class and self.default_child_resource_document:
    s_class = self._child_document_resources[self.default_child_resource_document]
```
